### ISSUE TITLE
VFS mount: tar command gets warning messages of "file changed as we read it" for unchanging files

### ISSUE THREAD

**Issue by leoluan1** [rclone/rclone#5277]
#### The associated forum post URL from `https://forum.rclone.org`

https://forum.rclone.org/t/rclone-mount-and-changing-timestamps-after-writes/22629/13?u=leoluan

#### What is the problem you are having with rclone?

Tar command caused a "file changed as we read it" warning message for every file when the files were not actually changed.

#### What is your rclone version (output from `rclone version`)

A development build before 1.55.0 was released, master branch at commit aea8776a4368ee8ab02b436717d219d0e868d701

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

Ubuntu 18.04.5 LTS

#### Which cloud storage system are you using? (e.g. Google Drive)

IBM COS (s3)

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)
```
root@tapsrv01:~# tar -cvf /spstg/test.tar /mnt/ibmcos-cache/clone/current/.git/config
tar: Removing leading `/' from member names
/mnt/ibmcos-cache/clone/current/.git/config
tar: /mnt/ibmcos-cache/clone/current/.git/config: file changed as we read it
root@tapsrv01:~#
```

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)
Here is the -vv log file from the test.
https://gist.github.com/leoluan1/e22f288e5bd7605cb422cfe58c2c5daf

**Comment by ncw** [comment]
I've attempted to fix this here

[v1.56.0-beta.5432.88eaae2f7.fix-5277-vfs-modtime-changing](https://beta.rclone.org/branch/fix-5277-vfs-modtime-changing/v1.56.0-beta.5432.88eaae2f7.fix-5277-vfs-modtime-changing/) on branch [fix-5277-vfs-modtime-changing](https://github.com/rclone/rclone/tree/fix-5277-vfs-modtime-changing) (uploaded in 15-30 mins)

What do you think - does that work for you? Patch look OK?

**Comment by leoluan1** [comment]
I retried the tests and this fix seems to work.   However, this fix reminds me of the question of how the item.metaDirty field is used for.  This field is set to true or false in multiple places and it's currently checked during the cache purge/item reset process.  Items with this field being true are not touched during purge/reset and item.   Do we really need to avoid purge/reset items with metaDirty being true?  If not, this field seems to be redundant and it causes confusion.

**Comment by leoluan1** [comment]
Reading the metaDirty related code, I guess maybe an intended real use of the item.metaDirty field is probably for the item._save() function call (or the IO within it) to be bypassed (currently not done)?  But it appears that whenever item._save() call is made, there is something changed in the item.Info field that we want to save the metadata file.   The checks of item.metaDIrty in item.inUse() may be redundant?  It looks like item.opens won't be zero if item.metaDirty is true. https://github.com/rclone/rclone/blob/5d799431a74568ac24c9b9023320fa7aa87b7d46/vfs/vfscache/item.go#L173

The check for item.metaDirty in reset does not seem necessary.  We would be removing the cache item anyway if the item satisfies other checks. 

Am I missing something for a true reason for having item.metaDirty.

**Comment by ncw** [comment]
> I retried the tests and this fix seems to work.

Great

> However, this fix reminds me of the question of how the item.metaDirty field is used for. This field is set to true or false in multiple places and it's currently checked during the cache purge/item reset process. Items with this field being true are not touched during purge/reset. Do we really need to avoid purge/reset items with metaDirty being true? If not, this field seems to be redundant and it causes confusion.

That is a good question. The purpose of `metaDirty` is to show that the metadata in memory is out of sync with that on disk and needs to be written back.

https://github.com/rclone/rclone/blob/08a2df51bea9b3bb71c463956a211e8dd37fb620/vfs/vfscache/item.go#L66

Specifically by metadata we mean the data that is written from the `Info` member to the `json` file.

So whenever we change the `Info` member we set `metaDirty = true` and at some time later we write that out to disk.

> Reading the metaDirty related code, I guess maybe an intended real use of the item.metaDirty field is for the item._save() function call (or the IO within it) to be bypassed (currently not done)? But it appears that whenever item._save() call is made, there is something changed in the item.Info field that we want to save it to the metadata file. The checks of item.metaDIrty in item.inUse() may be redundant? It looks like item.opens won't be zero if item.metaDirty is true.

Yes I see what you mean - we never avoid that IO.

All the uses of metaDirty actually seem to be wrong if it is just a flag to note whether the in memory data is out of date with respect to disk.

It looks like I got confused about the real purpose of metaDirty at some point.

> https://github.com/rclone/rclone/blob/5d799431a74568ac24c9b9023320fa7aa87b7d46/vfs/vfscache/item.go#L173
> 
> The check for item.metaDirty in reset does not seem necessary.
> 
> https://github.com/rclone/rclone/blob/5d799431a74568ac24c9b9023320fa7aa87b7d46/vfs/vfscache/item.go#L938
> 
> We would be removing the cache item anyway if the item satisfies other checks.
> Am I missing something for a true reason for having item.metaDirty.

I think we have two options

1. Just remove `metaDirty` completely

2. Remove all the comparisons for `metaDirty` (since they are all wrong) and add a new one in `_save` to not write the file if we don't have to. I'm not sure that would win us much though looking through the code noting where we call `_save`.

I'm probably thinking option 1. is the best - what do you think?

**Comment by leoluan1** [comment]
I agree. Option 1 seems better.  The item.metaDirty field contributes to a good deal of confusion when reading the code. :)
Since Close changes Atime and that needs to be written to the meta file, I don't see how ._save can avoid the meta file write.
https://github.com/rclone/rclone/blob/08a2df51bea9b3bb71c463956a211e8dd37fb620/vfs/vfscache/item.go#L641

**Comment by ncw** [comment]
OK I did that here

[v1.56.0-beta.5447.a0a0d6f60.fix-5277-vfs-modtime-changing](https://beta.rclone.org/branch/fix-5277-vfs-modtime-changing/v1.56.0-beta.5447.a0a0d6f60.fix-5277-vfs-modtime-changing/) on branch [fix-5277-vfs-modtime-changing](https://github.com/rclone/rclone/tree/fix-5277-vfs-modtime-changing) (uploaded in 15-30 mins)

I ran the VFS torture test on it - looks fine!

Look OK to you?

**Comment by leoluan1** [comment]
Yes, the fix looks great.  I have run a multi-hour test against the fix branch and it is working fine so far.

**Comment by ncw** [comment]
Thanks! I'll merge both those fixes now.

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.56
